mypy/types.py

```diff
@@ -561,37 +561,47 @@ class Instance(Type):
     The list of type variables may be empty.
     """
 
-    __slots__ = ('type', 'args', 'erased', 'invalid', 'type_ref')
+    __slots__ = ('type', 'args', 'erased', 'invalid', 'type_ref', 'final_value')
 
     def __init__(self, typ: mypy.nodes.TypeInfo, args: List[Type],
-                 line: int = -1, column: int = -1, erased: bool = False) -> None:
+                 line: int = -1, column: int = -1, erased: bool = False,
+                 final_value: Optional['LiteralType'] = None) -> None:
         super().__init__(line, column)
         self.type = typ
         self.args = args
         self.erased = erased  # True if result of type variable substitution
         self.invalid = False  # True if recovered after incorrect number of type arguments error
         self.type_ref = None  # type: Optional[str]
 
+        # The underlying Literal[...] value if this instance was created via a Final
+        # declaration. For example, if we did `x: Final = 3`, x would have an instance
+        # with a `final_value` of `LiteralType(3, int_fallback)`.
+        self.final_value = final_value
```

Comment:
I think this requires a (much) more extended comment. Adding a new field to instances is expensive, and although I understand why this is necessary, it is better to add:
* Short motivation explaining what we discussed in this PR.
* Large scale design (i.e. how this will be assigned in checker, and used by expression visitor to infer type)
* Maybe another example where this will be `None`.